### PR TITLE
Allocate primary shards based on allocation IDs

### PR DESCRIPTION
**Pull request by ywelsch**
- Add allocation IDs to `TransportNodesListGatewayStartedShards` action.
- Use the above to assign a primary shard on recovery.
- Also add allocation id to indices shard store response (`/some_index/_shard_stores`)

Relates to #14739

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 1]
```diff
@@ -71,12 +73,28 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
             }
 
             IndexMetaData indexMetaData = metaData.index(shard.getIndex());
-            Settings indexSettings = Settings.builder().put(settings).put(indexMetaData.getSettings()).build();
+            IndexSettings indexSettings = new IndexSettings(indexMetaData, settings, Collections.emptyList());
 
-            NodesAndVersions nodesAndVersions = buildNodesAndVersions(shard, recoverOnAnyNode(indexSettings), allocation.getIgnoreNodes(shard.shardId()), shardState);
-            logger.debug("[{}][{}] found {} allocations of {}, highest version: [{}]", shard.index(), shard.id(), nodesAndVersions.allocationsFound, shard, nodesAndVersions.highestVersion);
+            Set<String> allocationIds = indexMetaData.activeAllocationIds(shard.id());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: call this lastActiveAllocationIds?

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 2]
```diff
@@ -161,11 +243,11 @@ private boolean isEnoughAllocationsFound(ShardRouting shard, IndexMetaData index
     /**
      * Based on the nodes and versions, build the list of yes/no/throttle nodes that the shard applies to.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: comment should be - Split the list of nodes to lists of yes/no/throttle nodes based on allocation deciders (versions are now irrelevant)

**Comment by ywelsch** [comment]
Pushed changes related to our discussion:

- Use allocation ids and index creation version to determine `allocatedPostIndexCreate`
- Added integration test for #15241 and #14671

File: core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java [hunk 3]
```diff
@@ -115,9 +116,10 @@ private void writeTo(StreamOutput out) throws IOException {
         private StoreStatus() {
         }
 
-        public StoreStatus(DiscoveryNode node, long version, Allocation allocation, Throwable storeException) {
+        public StoreStatus(DiscoveryNode node, long version, String allocationId, Allocation allocation, Throwable storeException) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java]: That Allocation enum is really confusing in this context (I stumbled on this 3 times now). This is unrelated to this change, so I'm not suggesting we change it here - but do you have suggestions for a better name? maybe AllocatedAs ?

File: core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java [hunk 4]
```diff
@@ -213,7 +213,7 @@ void finish() {
              * A shard exists/existed in a node only if shard state file exists in the node
              */
             private boolean shardExistsInNode(final NodeGatewayStartedShards response) {
-                return response.storeException() != null || response.version() != -1;
+                return response.storeException() != null || response.version() != -1 || response.allocationId() != null;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java]: I'm not sure `response.allocationId() != null`  is OK. What happens when we allocate old indices that don't have allocationId serialized?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java [hunk 5]
```diff
@@ -95,7 +94,7 @@ public ClusterState execute(ClusterState currentState) {
                         IndexRoutingTable indexRoutingTable = currentState.routingTable().index(index);
                         for (IndexShardRoutingTable shard : indexRoutingTable) {
                             for (ShardRouting shardRouting : shard) {
-                                if (shardRouting.primary() == true && shardRouting.allocatedPostIndexCreate() == false) {
+                                if (shardRouting.primary() == true && shardRouting.allocatedPostIndexCreate(indexMetaData) == false) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java]: I think this protection can go away now? It's a protection against an index staying red when opening it (because now primary can be found). We now do a better job with the active allocation ids and we will just assign new empty primaries.

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 6]
```diff
@@ -58,7 +61,15 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
         while (unassignedIterator.hasNext()) {
             ShardRouting shard = unassignedIterator.next();
 
-            if (needToFindPrimaryCopy(shard) == false) {
+            if (shard.primary() == false) {
+                continue;
+            }
+
+            IndexMetaData indexMetaData = metaData.index(shard.getIndex());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: nit: make these final?

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 7]
```diff
@@ -70,13 +81,29 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
+
+            final NodesAndVersions nodesAndVersions;
+            final boolean enoughAllocationsFound;
+
+            if (lastActiveAllocationIds.isEmpty() && indexSettings.getIndexVersionCreated().before(Version.V_3_0_0)) {
+                // when we load an old index (after upgrading cluster) or restore a snapshot of an old index
+                // fall back to old version-based allocation mode
+                // Note that once the shard has been active, lastActiveAllocationIds will be non-empty
+                nodesAndVersions = buildNodesAndVersions(shard, recoverOnAnyNode(indexSettings), allocation.getIgnoreNodes(shard.shardId()), shardState);
+                logger.debug("[{}][{}] found {} allocations of {}, highest version: [{}]", shard.index(), shard.id(), nodesAndVersions.allocationsFound, shard, nodesAndVersions.highestVersion);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: can we add something about this being an old index to the log message?

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 8]
```diff
@@ -70,13 +81,29 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
 
-            IndexMetaData indexMetaData = metaData.index(shard.getIndex());
-            Settings indexSettings = Settings.builder().put(settings).put(indexMetaData.getSettings()).build();
+            final boolean snapshotRestore = shard.restoreSource() != null;
+            Set<String> lastActiveAllocationIds = indexMetaData.activeAllocationIds(shard.id());
+
+            final NodesAndVersions nodesAndVersions;
+            final boolean enoughAllocationsFound;
+
+            if (lastActiveAllocationIds.isEmpty() && indexSettings.getIndexVersionCreated().before(Version.V_3_0_0)) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: do we need the version check here? it's folded into allocatedPostIndexCreate() ? I think it will be simpler to read if we remove this from the if and add an assert on this, explaining why we expect it like that. Something like:

```
            if (lastActiveAllocationIds.isEmpty()) {
                assert indexSettings.getIndexVersionCreated().before(Version.V_3_0_0) : "trying to allocated a primary with an empty allocation id set, but index is new";
```

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 9]
```diff
@@ -70,13 +81,29 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
+                // when we load an old index (after upgrading cluster) or restore a snapshot of an old index
+                // fall back to old version-based allocation mode
+                // Note that once the shard has been active, lastActiveAllocationIds will be non-empty
+                nodesAndVersions = buildNodesAndVersions(shard, recoverOnAnyNode(indexSettings), allocation.getIgnoreNodes(shard.shardId()), shardState);
+                logger.debug("[{}][{}] found {} allocations of {}, highest version: [{}]", shard.index(), shard.id(), nodesAndVersions.allocationsFound, shard, nodesAndVersions.highestVersion);
 
-            NodesAndVersions nodesAndVersions = buildNodesAndVersions(shard, recoverOnAnyNode(indexSettings), allocation.getIgnoreNodes(shard.shardId()), shardState);
-            logger.debug("[{}][{}] found {} allocations of {}, highest version: [{}]", shard.index(), shard.id(), nodesAndVersions.allocationsFound, shard, nodesAndVersions.highestVersion);
+                enoughAllocationsFound = isEnoughAllocationsFound(shard, indexMetaData, nodesAndVersions);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: can we rename isEnoughAllocationsFound to isEnoughVersionBasedAllocationsFound?

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 10]
```diff
@@ -108,22 +135,71 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
+            }
 
-        // this is an API allocation, ignore since we know there is no data...
-        if (shard.allocatedPostIndexCreate() == false) {
-            return false;
+            if (nodeShardState.storeException() == null) {
+                if (allocationId == null) {
+                    logger.trace("[{}] on node [{}] has no allocation id in shard meta data", shard, nodeShardState.getNode());
+                } else {
+                    logger.trace("[{}] on node [{}] has allocation id [{}] of shard", shard, nodeShardState.getNode(), allocationId);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: I'm fine wil logging "allocation id [null] of shard" . Will save on the if :)

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java [hunk 11]
```diff
@@ -285,4 +285,11 @@ public void testMappingConflictRootCause() throws Exception {
             assertThat(messages.toString(), containsString("mapper [text] is used by multiple types"));
         }
     }
+
+    public void testRestartIndexCreationAfterFullClusterRestart() throws Exception {
+        client().admin().cluster().prepareUpdateSettings().setTransientSettings(Settings.builder().put("cluster.routing.allocation.enable", "none")).get();
+        client().admin().indices().prepareCreate("test").setSettings(indexSettings()).get();
+        internalCluster().fullRestart();
+        ensureGreen("test");
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java]: nice and clean :) another variant you may want to consider is creating an index with way more shards than nodes and see that we get to yellow (now we will wait for a quorum of copies)

File: core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+@ESIntegTestCase.SuppressLocalMode
+public class AllocationIdIT extends ESIntegTestCase {
+
+    public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception {
+        logger.info("--> starting 3 nodes, 1 master, 2 data");
+        // we are shutting down nodes - make sure we don't have 2 clusters if we test network
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java]: I don't get the comment?

File: core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        String master = internalCluster().startMasterOnlyNodeAsync().get();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java]: startMasterOnlyNode?

File: core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+@ESIntegTestCase.SuppressLocalMode
+public class AllocationIdIT extends ESIntegTestCase {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java]: Can we call this PrimaryAllocationIT ?

File: core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+        partition.stopDisrupting();
+
+        logger.info("--> waiting for node with old primary shard to rejoin the cluster");
+        ensureStableCluster(2, master);
+
+        logger.info("--> check that old primary shard does not get promoted to primary again");
+        // wait 5 seconds and check that we timed out
+        try {
+            client(master).admin().cluster()
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java]: it's shame the test times out every time (and will this be slow). How about the following sequence:

- stop distrupting
- wait for a two node cluster (like now).
- ensure that the index is still red (I know this is not bullet proof but if we do something wrong it will flag it enough times)
- start a new data node (which will reuse the old data folder)
- search and try to find the documents we just indexed.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 16]
```diff
@@ -59,25 +62,30 @@ public void buildTestAllocator() {
...
-    public void testNoProcessReplica() {
-        ShardRouting shard = TestShardRouting.newShardRouting("test", 0, null, null, null, false, ShardRoutingState.UNASSIGNED, 0, new UnassignedInfo(UnassignedInfo.Reason.CLUSTER_RECOVERED, null));
-        assertThat(testAllocator.needToFindPrimaryCopy(shard), equalTo(false));
-    }
-
-    public void testNoProcessPrimayNotAllcoatedBefore() {
-        ShardRouting shard = TestShardRouting.newShardRouting("test", 0, null, null, null, true, ShardRoutingState.UNASSIGNED, 0, new UnassignedInfo(UnassignedInfo.Reason.INDEX_CREATED, null));
-        assertThat(testAllocator.needToFindPrimaryCopy(shard), equalTo(false));
+    public void testNoProcessPrimaryNotAllocatedBefore() {
+        MetaData metaData = MetaData.builder()
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: maybe reuse the routingAllocationWithOnePrimaryNoReplicas method, but give a param whether it should see it as a recovery or as new? also, I think we want a variant where we are recovering, but the active allocation id is empty?

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 17]
```diff
@@ -97,11 +111,43 @@ public void testNoAllocationFound() {
...
+    }
+
+    /**
+     * Tests when the node returns that no data was found for it, it will be moved to ignore unassigned.
      */
     public void testStoreException() {
-        RoutingAllocation allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders());
-        testAllocator.addData(node1, 3, new CorruptIndexException("test", "test"));
+        final RoutingAllocation allocation;
+        if (randomBoolean()) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: we can one more legal state here - old index with allocation ids.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 18]
```diff
@@ -112,8 +158,14 @@ public void testStoreException() {
      * Tests that when there is a node to allocate the shard to, it will be allocated to it.
      */
     public void testFoundAllocationAndAllocating() {
-        RoutingAllocation allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders());
-        testAllocator.addData(node1, 10);
+        final RoutingAllocation allocation;
+        if (randomBoolean()) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: same comment about possible state.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 19]
```diff
@@ -126,8 +178,14 @@ public void testFoundAllocationAndAllocating() {
      * it will be moved to ignore unassigned until it can be allocated to.
      */
     public void testFoundAllocationButThrottlingDecider() {
-        RoutingAllocation allocation = routingAllocationWithOnePrimaryNoReplicas(throttleAllocationDeciders());
-        testAllocator.addData(node1, 10);
+        final RoutingAllocation allocation;
+        if (randomBoolean()) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: same...

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 20]
```diff
@@ -149,11 +213,11 @@ public void testFoundAllocationButNoDecider() {
     }
 
     /**
-     * Tests that the highest version node is chosed for allocation.
+     * Tests that the highest version node is chosen for allocation.
      */
     public void testAllocateToTheHighestVersion() {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: rename to testAllocateToTheHighestVersionOnLegacyIndex?

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 21]
```diff
@@ -190,7 +255,7 @@ public void testRestoreIgnoresNoNodesToAllocate() {
      */
     public void testEnoughCopiesFoundForAllocation() {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: same - add a OnLegacyIndex suffix?

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 22]
```diff
@@ -317,10 +410,39 @@ public void testAllocationOnAnyNodeShouldPutNodesWithExceptionsLast() {
...
+
+        nodes = testAllocator.buildAllocationIdBasedNodes(shard, false, true, ignoredNodes,
+                Collections.singleton("id1"), fetches).nodes;
+        assertThat(nodes.size(), equalTo(2));
+        // node3 should be last here
+        assertThat(nodes, contains(node1, node3));
+    }
+
+    /**
+     * if no activeAllocationIds are provided, we fallback to version-based mode
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: I don't follow the comment?

File: core/src/test/java/org/elasticsearch/gateway/QuorumGatewayIT.java [hunk 23]
```diff
@@ -51,72 +47,12 @@ protected int numberOfReplicas() {
         return 2;
     }
 
-    public void testChangeInitialShardsRecovery() throws Exception {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/QuorumGatewayIT.java]: can we deprecate the INDEX_RECOVERY_INITIAL_SHARDS setting? also add a breaking change note + remove docs for it? I'm fine with doing this in another PR, but we need to update the meta issue

**Comment by bleskes** [comment]
Thanks @ywelsch . I think this looks great. I left some comments and also want to ping @dakrone to discuss the recover on any node option. I hope we can get this simpler...

File: core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java (hunk 4 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java]: Note that the condition just requires any of the fields to have a value, so it will work fine in the case where we do not have allocation ids. We say that the shard exists if we get any sensible information about it. At the moment we have a version whenever we have an allocation id. But if we would remove versions in the future, that might not hold anymore. Adding `response.allocationId() != null` just adds a little more safety.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java (hunk 5 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java]: agree.

File: core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java]: you are right. Brain circuit break - interpreted the || as && . All good.

File: core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java (hunk 12 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java]: copy pasta from another test ;-) I removed it.

File: core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java (hunk 3 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java]: What about `AllocationStatus`? The JSON field should also be adapted.

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java (hunk 11 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java]: ok, will put this into PrimaryAllocationIT

File: core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java]: I liked allocatedAs better :) (note that the name of the java JSON field is ALLOCATED :))

File: core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java (hunk 15 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java]: I'll refine point 3 a bit:

- kick reroute
- wait for all shard states to be fetched (`assertBusy(() -> assertThat(internalCluster().getInstance(GatewayAllocator.class, master).getNumberOfInFlightFetch(), equalTo(0)));`)
- kick reroute a second time
- check that all shards are unassigned

**Comment by ywelsch** [comment]
@bleskes I pushed a new set of changes that address your comments.

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 24]
```diff
@@ -56,13 +60,26 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
-            AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> shardState = fetchData(shard, allocation);
+            final IndexMetaData indexMetaData = metaData.index(shard.getIndex());
+            final IndexSettings indexSettings = new IndexSettings(indexMetaData, settings, Collections.emptyList());
+
+            if (shard.allocatedPostIndexCreate(indexMetaData) == false) {
+                // when we create a fresh index
+                continue;
+            }
+
+            if (recoverOnAnyNode(indexSettings)) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: can we only skip allocation if we don't find any copy?

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 25]
```diff
@@ -108,23 +142,55 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
+                logger.trace("[{}] on node [{}] has allocation id [{}] of shard", shard, nodeShardState.getNode(), allocationId);
+            } else {
+                logger.trace("[{}] on node [{}] has allocation id [{}] but the store can not be opened, treating as no allocation id", nodeShardState.storeException(), shard, nodeShardState.getNode(), allocationId);
+                allocationId = null;
+            }
+
+            if (allocationId != null) {
+                if (lastActiveAllocationIds.contains(allocationId)) {
+                    matchingNodes.add(node);
+                    if (nodeShardState.version() > highestVersion) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: highestVersion  = Math.max(highestVersion, nodeShardState.version());  is faster and cleaner

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 26]
```diff
@@ -108,23 +142,55 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
     }
 
-    /**
-     * Does the shard need to find a primary copy?
-     */
-    boolean needToFindPrimaryCopy(ShardRouting shard) {
-        if (shard.primary() == false) {
-            return false;
-        }
+    protected NodesAndVersions buildAllocationIdBasedNodes(ShardRouting shard, boolean matchAnyAllocationId, Set<String> ignoreNodes,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: can we add java docs? I'm mostly after an explanation of matchAnyAllocationId

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 27]
```diff
@@ -159,13 +225,13 @@ private boolean isEnoughAllocationsFound(ShardRouting shard, IndexMetaData index
     }
 
     /**
-     * Based on the nodes and versions, build the list of yes/no/throttle nodes that the shard applies to.
+     * Split the list of nodes to lists of yes/no/throttle nodes based on allocation deciders
      */
-    private NodesToAllocate buildNodesToAllocate(ShardRouting shard, RoutingAllocation allocation, NodesAndVersions nodesAndVersions) {
+    NodesToAllocate buildNodesToAllocate(ShardRouting shard, RoutingAllocation allocation, List<DiscoveryNode> nodes) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: why was the private removed?

File: core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+
+        partition.stopDisrupting();
+
+        logger.info("--> waiting for node with old primary shard to rejoin the cluster");
+        ensureStableCluster(2, master);
+
+        logger.info("--> check that old primary shard does not get promoted to primary again");
+        // kick reroute and wait for all shard states to be fetched
+        client(master).admin().cluster().prepareReroute().get();
+        assertBusy(() -> assertThat(internalCluster().getInstance(GatewayAllocator.class, master).getNumberOfInFlightFetch(), equalTo(0)));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: nice one.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 17 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: ping?

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 18 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: ping

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 17 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: argh. Hidden by github ui. all good.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 18 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: same here :(

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 29]
```diff
@@ -59,25 +61,51 @@ public void buildTestAllocator() {
...
-     */
-    public void testNoProcessReplica() {
-        ShardRouting shard = TestShardRouting.newShardRouting("test", 0, null, null, null, false, ShardRoutingState.UNASSIGNED, 0, new UnassignedInfo(UnassignedInfo.Reason.CLUSTER_RECOVERED, null));
-        assertThat(testAllocator.needToFindPrimaryCopy(shard), equalTo(false));
+    public void testNoProcessPrimaryNotAllocatedBefore() {
+        final RoutingAllocation allocation;
+        if (randomBoolean()) {
+            allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders(), randomBoolean(), Version.CURRENT);
+        } else {
+            allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders(), true, Version.V_2_1_0);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: new is not possible with an older version...

**Comment by bleskes** [comment]
This looks great. Left some minor comment and one important one about the recover on any node settings. Also let's have another discussion with @dakrone about the failing test.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 29 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: Edge case: force allocation of a primary for a pre-3.0 index with reroute command.

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: I think we can work around it on the node side of things - it should expect to find an existing copy.

**Comment by ywelsch** [comment]
Pushed another set of changes, dealing with recover_on_any_node.

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 30]
```diff
@@ -70,25 +82,50 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
+            final boolean enoughAllocationsFound;
+
+            if (lastActiveAllocationIds.isEmpty()) {
+                assert indexSettings.getIndexVersionCreated().before(Version.V_3_0_0) : "trying to allocated a primary with an empty allocation id set, but index is new";
+                // when we load an old index (after upgrading cluster) or restore a snapshot of an old index
+                // fall back to old version-based allocation mode
+                // Note that once the shard has been active, lastActiveAllocationIds will be non-empty
+                nodesAndVersions = buildNodesAndVersions(shard, snapshotRestore || recoverOnAnyNode, allocation.getIgnoreNodes(shard.shardId()), shardState);
+                if (snapshotRestore || recoverOnAnyNode) {
+                    enoughAllocationsFound = nodesAndVersions.nodes.isEmpty() == false;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: nit: can we be consistent and use allocationsFound > 0?

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 31]
```diff
@@ -109,63 +146,99 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
...
+            String allocationId = nodeShardState.allocationId();
+
+            if (ignoreNodes.contains(node.id())) {
+                continue;
+            }
+
+            if (nodeShardState.storeException() == null) {
+                if (allocationId == null && nodeShardState.version() != -1) {
+                    // old shard with no allocation id, assign dummy so that it gets added below in case of matchAnyShard
+                    allocationId = "dummy";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: can we use "_n/a_" ?

File: docs/reference/migration/migrate_3_0.asciidoc [hunk 32]
```diff
@@ -472,3 +473,15 @@ from `OsStats.Cpu#getPercent`.
...
 Only stored fields are retrievable with this option.
 The fields option won't be able to load non stored fields from _source anymore.
+
+[[breaking_30_allocation]]
+=== Primary shard allocation
+
+Previously, primary shards were only assigned if a quorum of shard copies were found (configurable using
+`index.recovery.initial_shards`, now deprecated). In case where a primary had only a single replica, quorum was defined
+to be a single shard. This meant that any shard copy of an index with replication factor 1 could become primary, even it
+was a stale copy of the data on disk. This is fixed now by introducing allocation IDs.
```
Comment (bleskes) [inline docs/reference/migration/migrate_3_0.asciidoc]: This is now fixed by using allocation IDs. (the are already there)

File: docs/reference/modules/cluster/shards_allocation.asciidoc [hunk 33]
```diff
@@ -22,9 +22,8 @@ Enable or disable allocation for specific kinds of shards:
 
 This setting does not affect the recovery of local primary shards when
 restarting a node.  A restarted node that has a copy of an unassigned primary
-shard will recover that primary immediately, assuming that the
-<<index.recovery.initial_shards,`index.recovery.initial_shards`>> setting is
-satisfied.
+shard will recover that primary immediately, assuming that its allocation id matches
+one of the active allocation ids in the cluster.
```
Comment (bleskes) [inline docs/reference/modules/cluster/shards_allocation.asciidoc]: cluster *state*.

File: docs/reference/migration/migrate_3_0.asciidoc [hunk 34]
```diff
@@ -472,3 +473,15 @@ from `OsStats.Cpu#getPercent`.
...
+
+Allocation IDs assign unique identifiers to shard copies. This allows the cluster to differentiate between multiple
+copies of the same data and track which shards have been active, so that after a cluster restart, shard copies
+containing only the most recent data can become primaries.
```
Comment (bleskes) [inline docs/reference/migration/migrate_3_0.asciidoc]: I miss a note here about the change in recover on any node?

**Comment by bleskes** [comment]
LGTM. Left some extremely minor comments. No need for another review. Just merge after they are addressed. Thanks @ywelsch ! I can't tell you how happy I am for having this.
